// @ngx-gallery/core/esm2015/lib/components/templates/items.model.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { GalleryItemType } from '../../models/constants';
export class ImageItem {
    /**
     * @param {?} data
     */
    constructor(data) {
        this.data = data;
        this.type = GalleryItemType.Image;
    }
}
if (false) {
    /** @type {?} */
    ImageItem.prototype.type;
    /** @type {?} */
    ImageItem.prototype.data;
}
export class VideoItem {
    /**
     * @param {?} data
     */
    constructor(data) {
        this.data = data;
        this.type = GalleryItemType.Video;
    }
}
if (false) {
    /** @type {?} */
    VideoItem.prototype.type;
    /** @type {?} */
    VideoItem.prototype.data;
}
export class IframeItem {
    /**
     * @param {?} data
     */
    constructor(data) {
        this.data = data;
        this.type = GalleryItemType.Iframe;
    }
}
if (false) {
    /** @type {?} */
    IframeItem.prototype.type;
    /** @type {?} */
    IframeItem.prototype.data;
}
export class YoutubeItem {
    /**
     * @param {?} data
     */
    constructor(data) {
        this.data = {
            src: `//youtube.com/embed/${data.src}?wmode=transparent`,
            thumb: data.thumb ? data.thumb : `//img.youtube.com/vi/${data.src}/default.jpg`
        };
        this.type = GalleryItemType.Youtube;
    }
}
if (false) {
    /** @type {?} */
    YoutubeItem.prototype.type;
    /** @type {?} */
    YoutubeItem.prototype.data;
}
